modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,357 @@
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+{
+    firstName: String,
+    lastName*: String,
+    age?: Number
+}
+----
+
+==== Close and Open Objects
+
+Close object means that the type only accepts an object value if there are no additional fields in the object,
+just those key values specified by the type.
+
+Open object only put constraints on the fields that are declared in the type, and if they are all verified, then the type accepts the value. The type accepts objects that have other fields apart of the ones mentioned in the declaration. All object types are open unless specified.
+
+The following example works even if `age` is not mentioned in the type declaration because the type of the variable `user` is an open object:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+var user: {firstName: String, lastName: String} = {firstName: "John", lastName: "Smith", age: 34}
+----
+
+The syntax to specify a close object is `{| |}`. The following example, makes our user type only support the keys `firstName` and `lastName`, so this throws an exception because the `age` field is not accepted.
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+var user: {|firstName: String, lastName: String|} = {name: "John", lastName: "Smith", age: 34}
+----
+
+
+=== Function Type
+
+DataWeave is a functional language and functions are a first class citizen which means that functions are values
+and values have an associated type.
+
+The syntax to define a function type is:
+
+----
+(paramType: ParamType,...) -> ReturnType
+----
+
+For example, if you want to define in a constraint a function type that has a parameter of type String, a second parameter of type Number and return a Boolean, the correct syntax is:
+
+----
+(paramA: String, paramB: Number) -> Boolean
+----
+
+For example, you could define a function that receives another function as an argument:
+
+.DataWeave Script
+----
+%dw 2.0
+output application/json
+
+fun applyIDsChange(ids: Array<Number>, changeTo: (Number) -> Number): Array<Number> = ???
+----
+
+If we call the function `applyIDsChange` with some function that does not match the `changeTo` constraint `(Number) -> Number`, DataWeave throws a type checking error.
+For example, the following call works because `abs` is a function that takes a Number and returns a Number, which matches the `changeTo` parameter.
+
+----
+applyIDsChange([1,-6, 3, -8], abs)
+----
+
+But this call fails because `sum` is a function that takes an Array and returns a Number, so it does not match our parameter constraint.
+
+----
+applyIDsChange([1,-6, 3, -8], sum)
+----
+
+== Complex Types
+
+=== Any and Nothing
+
+In some cases you can not enforce a restriction as it accepts all the values. +
+
+* `Any` type accepts all possible values.
+* `Nothing` type accepts no value but it can be assigned to all the types. This type is not frequently used explicitly but it is by the type inference algorithm.
```

Comment:
```suggestion
* `Nothing` type accepts no value, but it can be assigned to all the types. This type is not frequently used explicitly, but it is used by the type inference algorithm.
```